File: 2026-06/2026-06-20/prime_factorization.py
def prime_factorization(n):
    factor_found = n

    for possible_factor in range(2, n // 2):
        if not n % possible_factor:
            factor_found = possible_factor
            break

    return [n] if factor_found == n else [factor_found] + prime_factorization(n // factor_found)

# --- TEST SUITE ---

tests_text = r'''
1. prime_factorization(20) should return [2, 2, 5].
2. prime_factorization(17) should return [17].
3. prime_factorization(15) should return [3, 5].
4. prime_factorization(35) should return [5, 7].
5. prime_factorization(999) should return [3, 3, 3, 37].
6. prime_factorization(360) should return [2, 2, 2, 3, 3, 5].
7. prime_factorization(510510) should return [2, 3, 5, 7, 11, 13, 17].
'''

import re

tests_regex = re.compile(r"(?P<number>\d+)\.\s(?P<function_call>.+) should return (?P<output>.+?)\.?$", re.MULTILINE)
test_data = [item.groupdict() for item in re.finditer(tests_regex, tests_text)]

def run_tests(test_data):
    print("——————————————————————————",
        "\n🧪 Starting Verification...",
        "\n——————————————————————————")

    fail_count = 0

    for test in test_data:
        function_call_output = eval(test['function_call'])
        test_output = eval(test['output'])
        
        if function_call_output == test_output:
            print(f"{test['number']}.✅ PASS - Function Call:\n{test['function_call']}")
        else:
            print(f"{test['number']}.❌ FAIL - Function Call:\n{test['function_call']}\nExpected:\n{test_output}\nGot:\n{function_call_output}")
            fail_count += 1

        print("————————————————————————————")

    print(
        f"⚠️ WARNING: {fail_count}/{len(test_data)} tests FAILED." if fail_count
        else "🎉 SUCCESS: All tests PASSED."
    )

run_tests(test_data)

File: 06 June/06-20/primeFactorization.js
function primeFactorization(n) {
/* Solution using while() loop
  const primeFactors = [];

  while (n !== 1) {
    let factorFound = n;
    
    for (let possibleFactor = 2; possibleFactor < n / 2; ++possibleFactor) {
      if (!(n % possibleFactor)) {
        factorFound = possibleFactor;
        break;
      }
    }

    primeFactors.push(factorFound);
    n = n / factorFound;
  }

  return primeFactors;
*/

// Solution using recursion
  let factorFound = n;

  for (let possibleFactor = 2; possibleFactor < n / 2; ++possibleFactor) {
    if (!(n % possibleFactor)) {
      factorFound = possibleFactor;
      break;
    }
  }

  return factorFound === n ? [n] : [factorFound].concat(primeFactorization(n / factorFound));
}

// --- TEST SUITE ---

const testsText = String.raw`
1. primeFactorization(20) should return [2, 2, 5].
2. primeFactorization(17) should return [17].
3. primeFactorization(15) should return [3, 5].
4. primeFactorization(35) should return [5, 7].
5. primeFactorization(999) should return [3, 3, 3, 37].
6. primeFactorization(360) should return [2, 2, 2, 3, 3, 5].
7. primeFactorization(510510) should return [2, 3, 5, 7, 11, 13, 17].
`;

const testsRegex = /(?<number>\d+)\.\s(?<functionCall>.+) should return (?<output>.+?)\.?$/gm;
const testData = [...testsText.matchAll(testsRegex).map(match => match.groups)];

function runTests(testData) {
  console.log("——————————————————————————",
            "\n🧪Starting Verification...",
            "\n——————————————————————————");

  let failCount = 0;
  
  testData.forEach(test => {
    const functionCallOutput = eval(test.functionCall);
    const testOutput = eval(`(${test.output})`);

    if (JSON.stringify(functionCallOutput) === JSON.stringify(testOutput)) {
      console.log(`${test.number}.✅PASS - Function Call: ${test.functionCall}`);
    } else {
      console.log(`${test.number}.❌FAIL - Function Call: ${test.functionCall}\nExpected: ${JSON.stringify(testOutput)}\nGot: ${JSON.stringify(functionCallOutput)}`);

      ++failCount;
    }
    console.log("————————————————————————————");
  })

  console.log(failCount
    ? `⚠️WARNING: ${failCount}/${testData.length} tests FAILED.`
    : "🎉SUCCESS: All tests PASSED."
  );
}

runTests(testData);